### PR TITLE
Allow zfs unshare <protocol> -a

### PR DESCRIPTION
**Pull request by loli10K**
I  ̶s̶t̶o̶l̶e̶ tried to complete the work done in https://github.com/zfsonlinux/zfs/pull/3238. I also wanted to add a new test script but i noticed all `zfs_share` and `zfs_unshare` tests are disabled. The man page wording can be probably improved.

### CODE REVIEW

**Comment by behlendorf** [comment]
@loli10K thanks for picking up this work.  We're you able to test the change locally to verify it works?  

We'd like to enable the `zfs_share` and `zfs_unshare` tests in the test suite but they're going to take a little work to get running on Linux.  The lack of test coverage is a shame and has help up making any major changes to the existing implementation.

**Comment by loli10K** [comment]
@behlendorf yes, it seems to work fine on my test box.

```
root@box:~# # setup(.ksh)
root@box:~# zpool create -f tank /var/tmp/zpool.dat
root@box:~# zfs create tank/fs1
root@box:~# zfs create tank/fs2
root@box:~# zfs create tank/fs3
root@box:~# zfs set sharenfs=on tank/fs1
sa_enable_share: share->sharepath=/tank/fs1, protocol=nfs
foreach_nfs_host_cb: key=rw, value=(null)
sharing *:/tank/fs1 with opts sec=sys,rw,no_subtree_check,no_root_squash,mountpoint
root@box:~# zfs set sharesmb=on tank/fs1
sa_enable_share: share->sharepath=/tank/fs1, protocol=nfs
foreach_nfs_host_cb: key=rw, value=(null)
sharing *:/tank/fs1 with opts sec=sys,rw,no_subtree_check,no_root_squash,mountpoint
sa_enable_share: share->sharepath=/tank/fs1, protocol=smb
root@box:~# zfs set sharenfs=on tank/fs2
sa_enable_share: share->sharepath=/tank/fs2, protocol=nfs
foreach_nfs_host_cb: key=rw, value=(null)
sharing *:/tank/fs2 with opts sec=sys,rw,no_subtree_check,no_root_squash,mountpoint
root@box:~# zfs set sharesmb=on tank/fs3
sa_enable_share: share->sharepath=/tank/fs3, protocol=smb
root@box:~# # share everything
root@box:~# zfs share -a && exportfs && net usershare list
/tank/fs2     	<world>
/tank/fs1     	<world>
tank_fs3
tank_fs1
root@box:~# # unshare only nfs
root@box:~# zfs unshare -a nfs && exportfs && net usershare list
sa_disable_share: share->sharepath=/tank/fs2, protocol=nfs
foreach_nfs_host_cb: key=rw, value=(null)
unsharing *:/tank/fs2
sa_disable_share: share->sharepath=/tank/fs1, protocol=nfs
foreach_nfs_host_cb: key=rw, value=(null)
unsharing *:/tank/fs1
tank_fs3
tank_fs1
root@box:~# # share everything again
root@box:~# zfs share -a && exportfs && net usershare list
sa_enable_share: share->sharepath=/tank/fs1, protocol=nfs
foreach_nfs_host_cb: key=rw, value=(null)
sharing *:/tank/fs1 with opts sec=sys,rw,no_subtree_check,no_root_squash,mountpoint
sa_enable_share: share->sharepath=/tank/fs1, protocol=smb
sa_enable_share: share->sharepath=/tank/fs2, protocol=nfs
foreach_nfs_host_cb: key=rw, value=(null)
sharing *:/tank/fs2 with opts sec=sys,rw,no_subtree_check,no_root_squash,mountpoint
/tank/fs2     	<world>
/tank/fs1     	<world>
tank_fs3
tank_fs1
root@box:~# # unshare only smb
root@box:~# zfs unshare -a smb && exportfs && net usershare list
sa_disable_share: share->sharepath=/tank/fs3, protocol=smb
sa_disable_share: share->sharepath=/tank/fs1, protocol=smb
/tank/fs2     	<world>
/tank/fs1     	<world>
root@box:~# 
```

I'm also willing to help enable those test, what kind of work would be required to pull this off?

**Comment by behlendorf** [comment]
> yes, it seems to work fine on my test box.

Great, I'll get this reviewed and tested too.

> I'm also willing to help enable those test, what kind of work would be required to pull this off?

Hopefully not too much at least for the basic functional tests.  It largely depends on if the test cases depends on Solaris specific utilities or behavior.  The first step would be to simply enable them and see what breaks first.  We'll also need to make sure the required nfs/samba packages get added to the buildbot.

**Comment by loli10K** [comment]
@behlendorf is `smb` even tested on other implementations? I tried digging into OpenZFS source repo and the only `smb` usage i could find in the zfs-tests directory is in `zfs_set/canmount_004_pos.ksh`.
On the other hand `nfs` is heavily (and not surprisingly) tested in both `zfs_share` and `zfs_unshare`.

To get things working we need to handle `is_shared` `not_shared` `unshare_fs` and `setup_nfs_server` functions in `include/libtest.shlib` (and probably add a new `setup_smb_server`?).
Also `$UNSHARE -F nfs` won't work on Linux, we would have to change the `cleanup` function in `zfs_unshare/zfs_unshare_00[1-3]_pos.ksh` test scripts.

I tried to enable both `zfs_share` and `zfs_unshare` tests only changing `include/libtest.shlib` to "support" some functions when we are running in a `is_linux` environment:

```
/usr/share/zfs/test-runner/bin/test-runner.py  -c /usr/share/zfs/runfiles/linux.run -i /usr/share/zfs/zfs-tests
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/setup (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_001_pos (run as root) [00:04] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_002_pos (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_003_pos (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_004_pos (run as root) [00:00] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_005_pos (run as root) [00:00] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_006_pos (run as root) [00:04] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_007_neg (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_008_neg (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_009_neg (run as root) [00:01] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_010_neg (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/zfs_share_011_pos (run as root) [00:00] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_share/cleanup (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_unshare/setup (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_unshare/zfs_unshare_001_pos (run as root) [00:03] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_unshare/zfs_unshare_002_pos (run as root) [00:00] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_unshare/zfs_unshare_003_pos (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_unshare/zfs_unshare_004_neg (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_unshare/zfs_unshare_005_neg (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zfs_unshare/cleanup (run as root) [00:00] [PASS]
```

**Comment by behlendorf** [comment]
> is smb even tested on other implementations? 

Good question.  If it's not the existing test suite then probably not.

Then step one will be to get the existing NFS testing working.  The Samba tests are less critical anyway since the samba server layers on the POSIX layer which is already extensively tested.  The NFS kernel server on the other hand requires some extra code kernel side which isn't exercised at all by the existing tests.

> I tried to enable both zfs_share and zfs_unshare tests 

Those initial results are pretty encouraging, 9/16 tests passing.  Feel free to make whatever modifications are need to get it working under Linux.  Illumos and Linux have very different NFS implementations so I'm not surprised things will need to be adjusted.

**Comment by loli10K** [comment]
I've added the test case but left it disabled until we get `zfs_share` and `zfs_unshare` working ... a side note on that, i was going to add a wrapper function around `$SHARE` and `$UNSHARE` since other platforms support multiple protocols in `share(1M)` (which if i understand correctly is itself a wrapper around filesystem specific tools) with `-F fstype` ... i don't think we can make it with a single command (unless we port `share` on Linux and wrap `/usr/sbin/exportfs` and `/usr/bin/net`).

**Comment by behlendorf** [comment]
Regarding `$SHARE` and `$UNSHARE` a wrapper function sounds like a good solution.  We don't want to start providing a very generically named `share(1M)` utility.

**Comment by loli10K** [comment]
@behlendorf can i just push all the code to enable those test in this PR? Or would it be better to handle it in a different PR?

**Comment by behlendorf** [comment]
I'd suggest adding it to this PR.

**Comment by behlendorf** [comment]
@loli10K you happened to push this while I was updating the buildbot so it wasn't added to the queue.  You'll need to quickly rebase this on the latest master and force update the branch to start the testing.

File: tests/zfs-tests/include/libtest.shlib [hunk 1]
```diff
@@ -1149,7 +1203,7 @@ function setup_nfs_server
 	fi
 
 	if is_linux; then
-		log_unsupported "Currently unsupported by the test framework"
+		log_note "NFS server is handled by the ZFS build bot, skipping."
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: How about: "NFS server must started prior to running test framework."

File: tests/zfs-tests/tests/functional/cli_root/zfs_unshare/zfs_unshare_006_pos.ksh [hunk 2]
```diff
@@ -44,9 +44,9 @@ verify_runnable "global"
 function cleanup
 {
 	log_must $ZFS unshare -a
+	log_must $ZFS destroy -f $TESTPOOL/$TESTFS/shared
 	log_must $ZFS destroy -f $TESTPOOL/$TESTFS/shared1
 	log_must $ZFS destroy -f $TESTPOOL/$TESTFS/shared2
-	log_must $ZFS destroy -f $TESTPOOL/$TESTFS/shared3
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zfs_unshare/zfs_unshare_006_pos.ksh]: Why the change?

Comment (loli10K) [inline tests/zfs-tests/tests/functional/cli_root/zfs_unshare/zfs_unshare_006_pos.ksh]: I had inadvertently used `$TESTDIR1, $TESTDIR2, $TESTDIR3` instead of `$TESTDIR $TESTDIR1, $TESTDIR2`, i changed also the datasets names to keep things consistent but forgot to replace all occurrences.

**Comment by loli10K** [comment]
@behlendorf i think i may have found a bug in the original (without this patch) libshare implementation for SMB: `zfs share` will gladly share any dataset with dots (`.`) in its name but its counterpart `zfs unshare` won't even try to unshare it:

```
root@box# net usershare list
root@box# 
root@box# strace -f -e trace=execve zfs share tank/testfs.24395/shared2
execve("/usr/sbin/zfs", ["zfs", "share", "tank/testfs.24395/shared2"], [/* 17 vars */]) = 0
Process 32508 attached
[pid 32508] execve("/usr/sbin/exportfs", ["exportfs", "-v"], [/* 17 vars */]) = 0
[pid 32508] +++ exited with 0 +++
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=32508, si_uid=0, si_status=0, si_utime=0, si_stime=0} ---
sa_enable_share: share->sharepath=/tank/testfs.24395/shared2, protocol=smb
Process 32509 attached
[pid 32509] execve("/usr/bin/net", ["/usr/bin/net", "-S", "127.0.0.1", "usershare", "add", "tank_testfs.24395_shared2", "/tank/testfs.24395/shared2", "Comment: /tank/testfs.24395/shar"..., "Everyone:F"], [/* 17 vars */]) = 0
Process 32511 attached
[pid 32511] +++ exited with 0 +++
[pid 32509] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=32511, si_uid=0, si_status=0, si_utime=0, si_stime=0} ---
[pid 32509] +++ exited with 0 +++
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=32509, si_uid=0, si_status=0, si_utime=2, si_stime=1} ---
+++ exited with 0 +++
root@box# 
root@box# net usershare list
tank_testfs.24395_shared2
root@box# 
root@box# strace -f -e trace=execve zfs unshare tank/testfs.24395/shared2
execve("/usr/sbin/zfs", ["zfs", "unshare", "tank/testfs.24395/shared2"], [/* 17 vars */]) = 0
Process 32521 attached
[pid 32521] execve("/usr/sbin/exportfs", ["exportfs", "-v"], [/* 17 vars */]) = 0
[pid 32521] +++ exited with 0 +++
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=32521, si_uid=0, si_status=0, si_utime=0, si_stime=0} ---
sa_disable_share: share->sharepath=/tank/testfs.24395/shared2, protocol=smb
+++ exited with 0 +++
root@box# 
root@box# net usershare list
tank_testfs.24395_shared2
root@box# 
```

For some reason `sa_disable_share` (or `smb_disable_share`) is dropping the ball along the way.

**Comment by loli10K** [comment]
> net usershare: usershares are currently disabled

I think we need to enable Samba user shares from the BB setup script ...

**Comment by behlendorf** [comment]
> I think we need to enable Samba user shares from the BB setup script ...

Can we just disable the Samba shares tests for this initial patch and just focus on NFS.

**Comment by loli10K** [comment]
@behlendorf that's ok with me, so far in this PR we're probably dealing with too much stuff already:

- zfs unshare [proto] -a implementation
- update ZTS to enable most of zfs_share/zfs_unshare (NFS) tests
- fix a bug in smb_is_share_active()

I will squash the first 3 commits and disable the new test script `zfs_unshare_006_pos.ksh` which is still failing because smb usershares.

**Comment by loli10K** [comment]
@behlendorf i don't think the failed test has anything to do with the changes in this PR, but anyway, is there some way i can re-run only that single one?

**Comment by dinatale2** [comment]
@loli10K @behlendorf I re-ran the Ubuntu 14.04 i686 TEST builder for this pull request even though it's not a required builder.

**Comment by loli10K** [comment]
@behlendorf sorry to bother you again, is there anything that has still to be addressed in this PR? I tried to keep my weekend free to work on remaining issues, maybe i can try to wrap this up.

**Comment by behlendorf** [comment]
@loli10K thanks for wrapping up all the loose ends on this one and sorry about the delay.  I was away all last week.  The updated patch LGTM and we can tackle the remaining Samba issues in a different PR.
